### ISSUE TITLE
Intransitivity in TypeAlias handling with `cast`

### ISSUE THREAD

**Issue by zwade** [microsoft/pyright#4598]
**Describe the bug**

The new handling of `TypeAlias` requirements introduces an intransitivity when working with `cast`-ed types.

**To Reproduce**

Consider the two following snippets of code.

```py
from typing import cast, TypeAlias

Obj: TypeAlias = cast(type[object], object)
x: Obj
```

and 

```py
from typing import cast, TypeAlias

_Obj = cast(type[object], object)
Obj: TypeAlias = _Obj
x: Obj
```

One would presume that both of these are equivalent (modulo beta-reduction I guess?) and that pyright would report the same set of errors on both. 

However, the former produces an error (`error: Invalid expression form for type alias definition`), whereas the latter does not. 

**Expected behavior**

I'm not totally sure what the respective PEPs say, but my desire would be for bot of these to be legal. While I recognize that expressions are not normally permitted for type aliases, in the case of a cast I would expect it to work (for context, we use this to retype SQLAlchemy tables' instances).

Regardless of whether that would be permitted, I would expect the two variants to have the same output.

**VS Code extension or command-line**

pyright (yarn, `1.1.293`)

**Comment by erictraut** [comment]
These should both be illegal but for different reasons. Type annotations do not support call expressions (so the first code sample is invalid) and cannot include variables (so the second code sample is invalid). I'll look at adding the missing error for the second code sample.

I don't know what your intent is here. If you want to define a type alias for `type[object]`, then you can simply use `Obj: TypeAlias = type[object]`. There's no need to do any casting.

**Comment by erictraut** [comment]
This will be addressed in the next release. Both of the above cases will now be properly flagged as illegal `TypeAlias` definitions, as they should be.

**Comment by zwade** [comment]
Thanks for taking a look at this! I suspected this would be the way things go &mdash; just wasn't sure how `cast` was supposed to play with TypeAliases

For context, this is how we were using it before, which is now invalid:

```py
class TableDefn:
  id = Column(Integer, ...)
  name = Column(Text, ...)

class TableInstType:
  id: int
  name: str

TableInst = cast(type[TableInstType], TableDefn)
```

Where (presumably because of the import bug fixed in 273) `TableInst` was valid as a type annotation

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.294, which I just published. It will also be included in a future release of pylance.
